### ISSUE TITLE
std::functional argument generates incorrect code

### ISSUE THREAD

**Issue by peterbygrave** [RosettaCommons/binder#41]
The following code fails to generate pybind11 code that can compile.

```
#ifndef THING_H
#define THING_H
#include <functional>
namespace work{
int a(std::function<int(double, int)> fn) {
  return fn(1.0, 2);
}
}
#endif
```

The errors are all similar to this:
```
std/functional.cpp:43:61: error: expected expression
                pybind11::class_<std::_Maybe_unary_or_binary_function<int,<const double &, int>>, std::shared_ptr<std::_Maybe_unary_or_binary_function<int,<double, int>>>, std::binary_function<co...
                                                                          ^
```

The problem is the second template argument to `std::_Maybe_unary_or_binary_function<int,<double, int>>` should not have the angle brackets around the arguments (it accepts variadic template). If I manually strip the brackets, the code compiles.

**Comment by lyskov** [comment]
I will look this up. In a meantime as a workaround please add `-class std::function` into Binder config to disable bindings generation for such types.

**Comment by v-yatsenko** [comment]
There are other issues with variadic template parameters related.
For example if we have a template function like following:
`
template <typename... T>
void foo(T... args);
`
And we use it with empty list of arguments:
`
foo()
`
binder will generate:
`
cl.def("foo", (void (*)()) foo<<>>, ....
`

**Comment by lyskov** [comment]
fixed in 18cc3539a79c118633c0e9459a1d2e10a755fdbc
